File: tests/js/test_scaling_history.mjs
import assert from "node:assert/strict";
import fs from "node:fs";
import vm from "node:vm";
import test from "node:test";
const window = {};
const elements = Object.fromEntries(["task-history", "task-history-search", "task-history-query", "task-history-prev", "task-history-next", "task-history-total", "task-history-page"].map(id => [id, {}]));
vm.runInNewContext(fs.readFileSync("static/js/scaling-history.js", "utf8"), {window, document:{getElementById:id => elements[id]}});
test("container specs, missing legacy values, and failure targets are truthful and escaped", () => {
  const html = window.ScalingHistory.render([{resource_id:"<script>", mode:"execute", status:"failed", error:"<bad>",
    before_spec:{replicas:3, containers:{app:{cpu_request_cores:0.5, memory_request_gb:1}, sidecar:{cpu_request_cores:0.1}}},
    target_spec:{replicas:2, containers:{app:{cpu_request_cores:0.25}}}}]);
  assert.match(html, /app \/ CPU Request/);
  assert.match(html, /sidecar/);
  assert.match(html, /0.25/);
  assert.match(html, /目标规格（未确认完成）/);
  assert.doesNotMatch(html, /<script>|<bad>/);
  const old = window.ScalingHistory.render([{mode:"execute", status:"success", target_spec:{cpu_cores:4}}]);
  assert.match(old, /未记录/);
  assert.match(old, /调配后规格（命令结果）/);
  const single = window.ScalingHistory.render([{before_spec:{containers:{app:{cpu_request_cores:1}}}, target_spec:{cpu_request_cores:0.5, memory_request_gb:2}}]);
  assert.match(single, /app \/ 内存 Request/);
});
test("loads all resources without a selected resource and paginates", async () => {
  const urls = [];
  const request = async url => { urls.push(url); return {items:[{resource_id:"vm-b"}], page:url.includes("page=2") ? 2 : 1, page_size:20, total:21}; };
  await window.ScalingHistory.load(request);
  assert.match(urls[0], /^\/api\/scaling-history\?/);
  assert.equal(elements["task-history-next"].disabled, false);
  elements["task-history-next"].onclick();
  await new Promise(resolve => setImmediate(resolve));
  assert.match(urls[1], /page=2/);
  assert.equal(elements["task-history-next"].disabled, true);
  elements["task-history-query"].value = "vm-a";
  elements["task-history-search"].onsubmit({preventDefault(){}});
  await new Promise(resolve => setImmediate(resolve));
  assert.match(urls[2], /page=1.*q=vm-a/);
});

File: tests/test_scaling_history.py
from flask import Flask

from resource_predict.api.scaling import register_scaling_routes
from resource_predict.services.scaling import tasks
from resource_predict.services.store.query import safe_int


def test_global_history_paginates_resources_and_preserves_snapshots(tmp_path, monkeypatch):
    monkeypatch.setattr(tasks, "TASKS_PATH", tmp_path / "scaling_tasks.json")
    monkeypatch.setattr(tasks.threading.Thread, "start", lambda self: None)
    resource = {"resource_id": "vm-a", "resource_type": "openstack_vm", "spec": {"cpu_cores": 4}}
    created = tasks.create_scaling_task(resource)
    resource["spec"]["cpu_cores"] = 16
    tasks._patch_task(created["task_id"], {"status": "success", "updated_at_ms": 100,
                                        "plan": {"target_spec": {"cpu_cores": 8}, "commands": ["private-command"]}})
    rows = tasks._read_tasks()
    rows[0]["created_at_ms"] = 10
    rows.append({"task_id": "legacy", "resource_id": "k8s:b", "created_at_ms": 20, "mode": "execute",
                 "status": "failed", "results": [{"stdout": "private-output"}]})
    tasks._write_tasks(rows)
    app = Flask(__name__)
    register_scaling_routes(app, {"get_resource_detail": lambda *_args, **_kwargs: None, "safe_int": safe_int})
    client = app.test_client()
    first = client.get("/api/scaling-history?page_size=1").get_json()
    assert first["total"] == 2
    assert first["items"][0]["resource_id"] == "k8s:b"
    assert first["items"][0]["before_spec"] is None
    second = client.get("/api/scaling-history?page_size=1&page=2").get_json()["items"][0]
    assert second["before_spec"] == {"cpu_cores": 4}
    assert second["target_spec"] == {"cpu_cores": 8}
    assert second["finished_at_ms"] == 100
    assert "commands" not in str(second) and "private-output" not in str(first)
    filtered = client.get("/api/scaling-history?q=VM-A").get_json()
    assert filtered["total"] == 1
    assert client.get("/api/scaling-history?page=-1&page_size=999").get_json()["page_size"] == 100
    assert client.get("/api/scaling-history?q=missing").get_json()["items"] == []
